cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -228,16 +278,25 @@ private void executeAction(ApplicationSpecification appSpec, WorkflowActionNode
         public void run() {
           ClassLoaders.setContextClassLoader(action.getClass().getClassLoader());
           try {
+            if (isCustomAction) {
```

Comment:
it can be moved to other method I think, like
```java
if(customAction) {
  runInTransaction(action);
}
else { 
  action.run();
}
```
And then `runInTransaction` does the job of wrapping the custom action's run in transaction.